examples/hello_world_util.h

```diff
@@ -195,93 +195,70 @@ void print_common_help()
     }
 }
 
-int server_connect(uint16_t server_port, uint8_t ai_family)
+int test_connect(const char *server, uint16_t server_port, sa_family_t af)
 {
-    struct addrinfo hints, *res, *t;
-    int lsock  = -1;
-    int dsock  = -1;
-    int optval = 1;
+    int sockfd   = -1;
+    int listenfd = -1;
+    int optval   = 1;
     char service[8];
+    struct addrinfo hints, *res, *t;
     int ret;
 
-    sprintf(service, "%u", server_port);
+    snprintf(service, 8, "%u", server_port);
     memset(&hints, 0, sizeof(struct addrinfo));
-    hints.ai_flags    = AI_PASSIVE;
-    hints.ai_family   = ai_family;
+    hints.ai_flags    = (server == NULL) ? AI_PASSIVE : 0;
+    hints.ai_family   = af;
     hints.ai_socktype = SOCK_STREAM;
 
     ret = getaddrinfo(NULL, service, &hints, &res);
     CHKERR_JUMP(ret < 0, "getaddrinfo() failed", out);
 
     for (t = res; t; t = t->ai_next) {
-        lsock = socket(t->ai_family, t->ai_socktype, t->ai_protocol);
-        if (lsock >= 0) {
-            ret = setsockopt(lsock, SOL_SOCKET, SO_REUSEADDR, &optval,
-                             sizeof(optval));
-            CHKERR_JUMP(ret < 0, "server setsockopt()", out_close_sock);
-
-            if (!bind(lsock, t->ai_addr, t->ai_addrlen)) {
-                break;
+        sockfd = socket(t->ai_family, t->ai_socktype, t->ai_protocol);
+        if (sockfd >= 0) {
+            if (server != NULL) {
+                if (!connect(sockfd, t->ai_addr, t->ai_addrlen)) {
+                    break;
+                }
+            } else {
+                ret = setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &optval,
+                                 sizeof(optval));
+                CHKERR_JUMP(ret < 0, "server setsockopt()", err_close_sockfd);
+
+                if (!bind(sockfd, t->ai_addr, t->ai_addrlen)) {
+                    listenfd = sockfd;
+                    sockfd = -1;
+                    ret = listen(listenfd, 0);
+                    CHKERR_JUMP(ret < 0, "listen server", out_close_lfd);
+
+                    /* Accept next connection */
+                    fprintf(stdout, "Waiting for connection...\n");
+                    sockfd = accept(listenfd, NULL, NULL);
+                    CHKERR_JUMP(sockfd < 0, "accept server", out_close_lfd);
```

Comment:
better handle connection accept outside of the common function. if we managed to bind, should be ok to use this address entry.